### ISSUE TITLE
Middleware still causes extra queries

### ISSUE THREAD

**Issue by vhtkrk** [kolofordjango/kolo#4]
Hey,

Thanks for the release of 1.1.1, it seems to have *mostly* fixed the issue of the middleware causing extra queries to be evaluated, and I'm now a step closer to enjoying a smooth debug experience. Alas, I still found a corner case where Kolo causes the queries to bubble up with `django-rest-framework` when you combine any sort of related field and a SerializerMethodField.

Here is a minimal reproducible example:

```python
# models.py
from django.db import models

class A(models.Model):
    slug = models.SlugField()

class B(models.Model):
    a = models.ForeignKey(A, on_delete=models.CASCADE)
```
```python
# serializers.py
from rest_framework import serializers
from .models import A, B

class BSerializer(serializers.ModelSerializer):
    a = serializers.SlugRelatedField(slug_field='slug', queryset=A.objects.all())
    c = serializers.SerializerMethodField()

    def get_c(self, b):
        return []

    class Meta:
        model = B
        fields = ('a', 'c',)
```
```python
# viewsets.py
from rest_framework import viewsets
from .models import A, B
from .serializers import BSerializer

class BViewSet(viewsets.ModelViewSet):
    queryset = B.objects.all().select_related('a')
    serializer_class = BSerializer

```
```python
# urls.py
from django.contrib import admin
from django.conf.urls import include
from django.urls import path
from rest_framework import routers
from app.viewsets import BViewSet

router = routers.DefaultRouter()
router.register('b', BViewSet, 'b')

urlpatterns = [
    path('admin/', admin.site.urls),
    path('', include(router.urls)),
]
```

### Steps:
1. Setup a bunch of `A`s and `B`s in django shell:
```python
>>> import random, string
>>> from app.models import A, B
>>> # Create 100 A's in bulk
>>> As = [A(slug=''.join(random.choice(string.ascii_lowercase) for _ in range(8))) for _ in range(100)]
>>> A.objects.bulk_create(As)
>>> # Refresh A's from database
>>> As = list(A.objects.all())
>>> # Create 30 B's with randomly chosen A relations
>>> Bs = [B(a=random.choice(As)) for _ in range(30)]
>>> B.objects.bulk_create(Bs)
```

2. Run `manage.py runserver` and `curl http://localhost:8000/b/`

Without Kolo middleware, this does 1 SQL query. With the middleware, it does 61.

This is repeated 60 times:
```SQL
SELECT
  "app_a"."id",
  "app_a"."slug"
FROM
  "app_a"
LIMIT
  21
```

This seems to happen because `get_c` is called for every `B`, and the stack frame with the locals is saved. Locals obviously contains `self` which is a `BSerializer` whose `repr()` contains a list of all the fields, and the `a` field's `repr()` contains the queryset which will be evaluated even though it shouldn't.

**Comment by Ian-Foote** [comment]
Hmm, I'm not sure this is going to be easy to find a clean fix for.

Essentially, whenever some code (in this case a rest framework field) has a `__repr__`  that delegates to `repr(queryset)` we'll generate extra queries.

I'm not sure we have a good way to detect such a `__repr__` in general, which just leaves special-casing rest framework fields and any other library code doing a similar thing.

**Comment by vhtkrk** [comment]
I peeked into the middleware code, and got an idea from your current workaround of the problem.

You can fix this by monkey-patching Django's `QuerySet.__repr__()` to: 
```python
    def __repr__(self):
        if self._result_cache is None:
            return f"Unevaluated queryset for: {self.model.__name__}"
        data = list(self[:REPR_OUTPUT_SIZE + 1])
        if len(data) > REPR_OUTPUT_SIZE:
            data[-1] = "...(remaining elements truncated)..."
        return '<%s %r>' % (self.__class__.__name__, data)
```

I'm not sure whether this change will have some other side effects, but couldn't spot any in my quick tests (trying a bunch of views and running our app's suite of unit tests).

**Comment by vhtkrk** [comment]
Yeah, I can see your point. Off the top of my head it could affect logging stuff. I'm not sure I can think of any others, as it doesn't seem that any magic in Django itself relies on this behavior.

On the other hand, a debugger that spuriously changes the behavior of the program it's supposed to profile is also kind of self-defeating...

I just got another idea, how about monkey-patching django just for the duration of the serialization in `profiler.py`?

```python
from contextlib import contextmanager

def _qs_repr(self):
    if self._result_cache is None:
        return f"Unevaluated queryset for: {self.model.__name__}"
    data = list(self[:20 + 1])
    if len(data) > 20:
        data[-1] = "...(remaining elements truncated)..."
    return '<%s %r>' % (self.__class__.__name__, data)

_original_qs_repr = models.QuerySet.__repr__

@contextmanager
def patched_django():
    models.QuerySet.__repr__ = _qs_repr
    yield
    models.QuerySet.__repr__ = _original_qs_repr

def serialize_locals(locals: Dict[str, object]) -> str:
    with patched_django():
        serialized_locals = {key: _serialize_local(value) for key, value in locals.items()}
    return json.dumps(serialized_locals, indent=2)
```

Seems to work and won't affect code written by the users.

**Comment by Ian-Foote** [comment]
I've ended up doing something very similar to this suggestion. The main difference is I've decided to monkeypatch in `KoloProfiler.__call__` - that way we only toggle the monkeypatch when we enter/exit the profiling code.

**Comment by Ian-Foote** [comment]
This should be fixed in the next release.

**Comment by wilhelmklopp** [comment]
Just released 1.1.2 which includes this fix! 🎉  Thanks for implementing this so quickly @Ian-Foote 

And thanks so much for the detailed write up and for the great idea of the temporary monkey patch ✨ @vhtkrk 

Looking forward to seeing if you run into anything else or if it's smooth sailing from here on out! Also excited to hear how useful Kolo will be for you in practice and what we can do to make it even more useful! 😍
